server/src/main/java/org/opensearch/action/search/TransportSearchAction.java

```diff
@@ -490,11 +479,12 @@ private void executeRequest(
             relativeStartNanos,
             System::nanoTime
         );
-
-        final List<SearchRequestOperationsListener> searchListenersList = createSearchListenerList(originalSearchRequest, timeProvider);
-        SearchRequestContext searchRequestContext = new SearchRequestContext(
-            new SearchRequestOperationsListener.CompositeListener(searchListenersList, logger)
+        timeProvider.setEnabled(clusterService.getClusterSettings().get(SEARCH_PHASE_TOOK_ENABLED), originalSearchRequest);
```

Comment:
This strikes me as weird. We're setting the enabled flag here then passing it to a function that includes this based on the value of this flag. Makes me think the construction of the composite listener might better belong here?

Reply:
Besides `TimeProvider`, there might be other "per-request" listeners need to be added from other component to the `compositeListeners`, thus I think it would be better to offload the logic to a separate `listener manager` (or `SearchRequestOperationsListeners`), otherwise it would clutter the core search path code, and potentially resulting in a monolithic core path that may hinder future migration to a distributed service architecture.

Also the `timeprovider` is a special listener that reused some previous existing class. In an idea world, it should be also added from outside of core search path to avoid cluttering the core search path. But we can make follow up PRs to refactor this as well.

Reply:
> we can make follow up PRs to refactor this as well.

`TimeProvider` is interesting and definitely relevant here. This suggests the current API of the base class isn't quire right. Maybe the abstract class should expose a get enabled method with the signature `boolean getEnabled(SearchRequest)` to support choosing whether it is enabled on a per-request basis?

Reply:
@andrross Thanks for the input, and that's actually a good point! But for `timeProvider`, "whether it is enabled or not" is decided by both the clusterSetting level (`SEARCH_PHASE_TOOK_ENABLED`), and per request level. And that's also the reason I wrote this `void setEnabled(boolean enabledAtClusterLevel, SearchRequest searchRequest)` function to decide whether we should enable per phase time information tracking.  (https://github.com/opensearch-project/OpenSearch/pull/11526/files#diff-8bd5d5cc44233ae5dec63a6c7d414eb3ecaa6c72c8a5a1aee8f97b57e0c34142R338)

We can also move this part of logic to a getter, `boolean getEnabled(SearchRequest)`, but in that case we need to encapsolate the clusterSetting variable into the `timeProvider` class, otherwise it would be similar to the current approach. 

What do you think?

Reply:
Thanks for the feedback folks. The previous `TimeProvider` implementation might not be optimal. Let me actually update the PR to further refactor this to remove the `TimeProvider` since we already have the that info in `SeachRequestContext`. In that way the logic here would be much more clear.